File: qt/multimedia/gen_qvideosink.go
package multimedia

import (
	"unsafe"

	"github.com/mappu/miqt/qt"
)

type QVideoSink struct {
	h          uintptr
	isSubclass bool
}

// NewQVideoSink constructs a new QVideoSink object.
func NewQVideoSink() *QVideoSink {
	ret := newQVideoSink(QVideoSink_new())
	ret.isSubclass = true
	return ret
}

// NewQVideoSink2 constructs a new QVideoSink object.
func NewQVideoSink2(parent *qt.QObject) *QVideoSink {
	ret := newQVideoSink(QVideoSink_new2((*QObject)(parent.UnsafePointer())))
	ret.isSubclass = true
	return ret
}

func (this *QVideoSink) MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QVideoSink_MetaObject(this.h)))
}

func (this *QVideoSink) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QVideoSink_Metacast(this.h, param1_Cstring))
}

func QVideoSink_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QVideoSink_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QVideoSink) VideoSize() *qt.QSize {
	_goptr := qt.UnsafeNewQSize(unsafe.Pointer(QVideoSink_VideoSize(this.h)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QVideoSink) SubtitleText() string {
	var _ms struct_miqt_string = QVideoSink_SubtitleText(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QVideoSink) SetSubtitleText(subtitle string) {
	subtitle_ms := struct_miqt_string{}
	subtitle_ms.data = CString(subtitle)
	subtitle_ms.len = size_t(len(subtitle))
	defer free(unsafe.Pointer(subtitle_ms.data))
	QVideoSink_SetSubtitleText(this.h, subtitle_ms)
}

func (this *QVideoSink) SetVideoFrame(frame *QVideoFrame) {
	QVideoSink_SetVideoFrame(this.h, frame.cPointer())
}

func (this *QVideoSink) VideoFrame() *QVideoFrame {
	_goptr := newQVideoFrame(QVideoSink_VideoFrame(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QVideoSink) VideoFrameChanged(frame *QVideoFrame) {
	QVideoSink_VideoFrameChanged(this.h, frame.cPointer())
}

func (this *QVideoSink) OnVideoFrameChanged(slot func(frame *QVideoFrame)) {
	QVideoSink_connect_VideoFrameChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QVideoSink_VideoFrameChanged
func miqt_exec_callback_QVideoSink_VideoFrameChanged(cb intptr_t, frame *QVideoFrame) {
	gofunc, ok := cgo.Handle(cb).Value().(func(frame *QVideoFrame))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQVideoFrame(frame)

	gofunc(slotval1)
}

func (this *QVideoSink) SubtitleTextChanged(subtitleText string) {
	subtitleText_ms := struct_miqt_string{}
	subtitleText_ms.data = CString(subtitleText)
	subtitleText_ms.len = size_t(len(subtitleText))
	defer free(unsafe.Pointer(subtitleText_ms.data))
	QVideoSink_SubtitleTextChanged(this.h, subtitleText_ms)
}

func (this *QVideoSink) OnSubtitleTextChanged(slot func(subtitleText string)) {
	QVideoSink_connect_SubtitleTextChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QVideoSink_SubtitleTextChanged
func miqt_exec_callback_QVideoSink_SubtitleTextChanged(cb intptr_t, subtitleText struct_miqt_string) {
	gofunc, ok := cgo.Handle(cb).Value().(func(subtitleText string))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	var subtitleText_ms struct_miqt_string = subtitleText
	subtitleText_ret := GoStringN(subtitleText_ms.data, int(int64(subtitleText_ms.len)))
	free(unsafe.Pointer(subtitleText_ms.data))
	slotval1 := subtitleText_ret

	gofunc(slotval1)
}

func (this *QVideoSink) VideoSizeChanged() {
	QVideoSink_VideoSizeChanged(this.h)
}

func (this *QVideoSink) OnVideoSizeChanged(slot func()) {
	QVideoSink_connect_VideoSizeChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QVideoSink_VideoSizeChanged
func miqt_exec_callback_QVideoSink_VideoSizeChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func QVideoSink_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QVideoSink_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QVideoSink_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QVideoSink_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QVideoSink) callVirtualBase_MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QVideoSink_virtualbase_MetaObject(unsafe.Pointer(this.h))))
}

func (this *QVideoSink) OnMetaObject(slot func(super func() *qt.QMetaObject) *qt.QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QVideoSink_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QVideoSink_MetaObject
func miqt_exec_callback_QVideoSink_MetaObject(self QVideoSink, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *qt.QMetaObject) *qt.QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QVideoSink{h: self}).callVirtualBase_MetaObject)

	return (*QMetaObject)(virtualReturn.UnsafePointer())
}

func (this *QVideoSink) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QVideoSink_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QVideoSink) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QVideoSink_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QVideoSink_Metacast
func miqt_exec_callback_QVideoSink_Metacast(self QVideoSink, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QVideoSink{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/multimedia/gen_qvideosink.cpp
// +build ignore

#include <QMetaObject>
#include <QObject>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVideoFrame>
#include <QVideoSink>
#include <qvideosink.h>
#include "gen_qvideosink.h"
class MiqtVirtualQVideoSink : public virtual QVideoSink {
public:
MiqtVirtualQVideoSink(): QVideoSink() {};
MiqtVirtualQVideoSink(QObject* parent): QVideoSink(parent) {};

virtual ~MiqtVirtualQVideoSink() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QVideoSink::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QVideoSink_MetaObject(const_cast<MiqtVirtualQVideoSink*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QVideoSink::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QVideoSink::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QVideoSink_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QVideoSink::qt_metacast(param1);

	}
};
QVideoSink* QVideoSink_new() {
return new MiqtVirtualQVideoSink();
}
QVideoSink* QVideoSink_new2(QObject* parent) {
return new MiqtVirtualQVideoSink(parent);
}
void QVideoSink_virtbase(QVideoSink* src
, QObject** outptr_QObject
) {
*outptr_QObject = static_cast<QObject*>(src);
}
QMetaObject* QVideoSink_MetaObject(const QVideoSink* self) {
	return (QMetaObject*) self->metaObject();
}
void* QVideoSink_Metacast(QVideoSink* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QVideoSink_Tr(const char* s) {
	QString _ret = QVideoSink::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
QSize* QVideoSink_VideoSize(const QVideoSink* self) {
	return new QSize(self->videoSize());
}
struct miqt_string QVideoSink_SubtitleText(const QVideoSink* self) {
	QString _ret = self->subtitleText();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QVideoSink_SetSubtitleText(QVideoSink* self, struct miqt_string subtitle) {
	QString subtitle_QString = QString::fromUtf8(subtitle.data, subtitle.len);
	self->setSubtitleText(subtitle_QString);
}
void QVideoSink_SetVideoFrame(QVideoSink* self, QVideoFrame* frame) {
	self->setVideoFrame(*frame);
}
QVideoFrame* QVideoSink_VideoFrame(const QVideoSink* self) {
	return new QVideoFrame(self->videoFrame());
}
void QVideoSink_VideoFrameChanged(const QVideoSink* self, QVideoFrame* frame) {
	self->videoFrameChanged(*frame);
}
void QVideoSink_connect_VideoFrameChanged(QVideoSink* self, intptr_t slot) {
	MiqtVirtualQVideoSink::connect(self, static_cast<void (QVideoSink::*)(const QVideoFrame&) const>(&QVideoSink::videoFrameChanged), self, [=](const QVideoFrame& frame) {
		const QVideoFrame& frame_ret = frame;
		// Cast returned reference into pointer
		QVideoFrame* sigval1 = const_cast<QVideoFrame*>(&frame_ret);
		miqt_exec_callback_QVideoSink_VideoFrameChanged(slot, sigval1);
	});
}
void QVideoSink_SubtitleTextChanged(const QVideoSink* self, struct miqt_string subtitleText) {
	QString subtitleText_QString = QString::fromUtf8(subtitleText.data, subtitleText.len);
	self->subtitleTextChanged(subtitleText_QString);
}
void QVideoSink_connect_SubtitleTextChanged(QVideoSink* self, intptr_t slot) {
	MiqtVirtualQVideoSink::connect(self, static_cast<void (QVideoSink::*)(const QString&) const>(&QVideoSink::subtitleTextChanged), self, [=](const QString& subtitleText) {
		const QString subtitleText_ret = subtitleText;
		// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
		QByteArray subtitleText_b = subtitleText_ret.toUtf8();
		struct miqt_string subtitleText_ms;
		subtitleText_ms.len = subtitleText_b.length();
		subtitleText_ms.data = static_cast<char*>(malloc(subtitleText_ms.len));
		memcpy(subtitleText_ms.data, subtitleText_b.data(), subtitleText_ms.len);
		struct miqt_string sigval1 = subtitleText_ms;
		miqt_exec_callback_QVideoSink_SubtitleTextChanged(slot, sigval1);
	});
}
void QVideoSink_VideoSizeChanged(QVideoSink* self) {
	self->videoSizeChanged();
}
void QVideoSink_connect_VideoSizeChanged(QVideoSink* self, intptr_t slot) {
	MiqtVirtualQVideoSink::connect(self, static_cast<void (QVideoSink::*)()>(&QVideoSink::videoSizeChanged), self, [=]() {
		miqt_exec_callback_QVideoSink_VideoSizeChanged(slot);
	});
}
struct miqt_string QVideoSink_Tr2(const char* s, const char* c) {
	QString _ret = QVideoSink::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QVideoSink_Tr3(const char* s, const char* c, int n) {
	QString _ret = QVideoSink::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QVideoSink_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQVideoSink*>( (QVideoSink*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QVideoSink_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQVideoSink*)(self) )->virtualbase_MetaObject();
}
void QVideoSink_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQVideoSink*>( (QVideoSink*)(self) )->handle__Metacast = slot;
}
void* QVideoSink_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQVideoSink*)(self) )->virtualbase_Metacast(param1);
}
void QVideoSink_Delete(QVideoSink* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQVideoSink*>( self );
	} else {
		delete self;
	}
}